File: 程序员面试金典/2.5链表求和/addTwoNumbers_2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 递归版本
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addList(l1, l2, 0);
    }

    private ListNode addList(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null && carry == 0)
            return null;
        int x = l1 != null ? l1.val : 0;
        int y = l2 != null ? l2.val : 0;
        int s = x + y + carry;
        ListNode result = new ListNode(s % 10);
        result.next = addList(l1 != null ? l1.next : null, l2 != null ? l2.next : null, s / 10);
        return result;
    }

}

File: Algorithms/445/addTwoNumbers_2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# 方法2：转换为数字
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        return num2list(list2num(l1) + list2num(l2))


def list2num(head):
    res = 0
    while head:
        res = 10 * res + head.val
        head = head.next
    return res


def num2list(num):
    res = ListNode(0)
    while num:
        res.next = ListNode(num % 10, res.next)
        num //= 10
    return res.next or res
